lib/battle/CObstacleInstance.cpp

```diff
@@ -66,73 +69,140 @@ std::vector<BattleHex> CObstacleInstance::getAffectedTiles() const
 	}
 }
 
-// bool CObstacleInstance::spellGenerated() const
-// {
-// 	if(obstacleType == USUAL || obstacleType == ABSOLUTE_OBSTACLE)
-// 		return false;
-//
-// 	return true;
-// }
-
 bool CObstacleInstance::visibleForSide(ui8 side, bool hasNativeStack) const
 {
 	//by default obstacle is visible for everyone
 	return true;
 }
 
+int CObstacleInstance::getAnimationYOffset(int imageHeight) const
+{
+	int offset = imageHeight % 42;
+	if(obstacleType == CObstacleInstance::USUAL)
+	{
+		if(getInfo().blockedTiles.front() < 0 || offset > 37) //second or part is for holy ground ID=62,65,63
+			offset -= 42;
+	}
+	return offset;
+}
+
 bool CObstacleInstance::stopsMovement() const
 {
-	return obstacleType == QUICKSAND || obstacleType == MOAT;
+	return obstacleType == MOAT;
 }
 
 bool CObstacleInstance::blocksTiles() const
 {
-	return obstacleType == USUAL || obstacleType == ABSOLUTE_OBSTACLE || obstacleType == FORCE_FIELD;
+	return obstacleType == USUAL || obstacleType == ABSOLUTE_OBSTACLE ;
+}
+
+bool CObstacleInstance::triggersEffects() const
+{
+	return false;
 }
 
 SpellCreatedObstacle::SpellCreatedObstacle()
 {
-	casterSide = -1;
-	spellLevel = -1;
-	casterSpellPower = -1;
+	casterSide = 0;
+	spellLevel = 0;
+	casterSpellPower = 0;
 	turnsRemaining = -1;
-	visibleForAnotherSide = -1;
+	hidden = false;
+	passable = false;
+	trigger = false;
+	trap = false;
+	animationYOffset = 0;
+	revealed = false;
 }
 
 bool SpellCreatedObstacle::visibleForSide(ui8 side, bool hasNativeStack) const
 {
-	switch(obstacleType)
+	//we hide mines and not discovered quicksands
+	//quicksands are visible to the caster or if owned unit stepped into that particular patch
+	//additionally if side has a native unit, mines/quicksands will be visible
+
+	return casterSide == side || !hidden || revealed || hasNativeStack;
+}
+
+bool SpellCreatedObstacle::blocksTiles() const
+{
+	if(obstacleType == SPELL_CREATED)
+		return !passable;
+	else
+		return CObstacleInstance::blocksTiles();
+}
+
+bool SpellCreatedObstacle::stopsMovement() const
+{
+	if(obstacleType == SPELL_CREATED)
+		return trap;
+	else
+		return CObstacleInstance::stopsMovement();
+}
+
+bool SpellCreatedObstacle::triggersEffects() const
+{
+	if(obstacleType == SPELL_CREATED)
+		return trigger;
+	else
+		return CObstacleInstance::triggersEffects();
+}
+
+void SpellCreatedObstacle::toInfo(ObstacleChanges & info)
+{
+	info.id = uniqueID;
+	info.operation = ObstacleChanges::EOperation::ADD;
+
+	info.data.clear();
+	JsonSerializer ser(nullptr, info.data);
+	ser.serializeStruct("obstacle", *this);
+}
+
+void SpellCreatedObstacle::fromInfo(const ObstacleChanges & info)
+{
+	uniqueID = info.id;
+	obstacleType = SPELL_CREATED;

```

Comment:
I think that it is better way to initialize "obstacleType" in header or constructor instead here.